File: services/action-engine/src/nova_action_engine/api/approvals.py
"""`POST /v1/action/approvals/{id}/decide` -- the stopgap decision
endpoint (TDD 3D §4 point 4), served directly by this engine's own FastAPI
app since `api-gateway` does not exist yet
(`docs/design/phase-3/03-gateway-web-prerequisite.md`). Named deliberately
to mirror doc 11's reserved `/v1/autonomy/approvals/{id}/decide` naming
convention, under `action-engine`'s own namespace instead."""

from __future__ import annotations

from datetime import UTC, datetime
from uuid import UUID

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

router = APIRouter(prefix="/v1/action", tags=["approvals"])


class PendingApprovalResponse(BaseModel):
    """One approval still waiting on a human, as the Approvals panel reads it."""

    action_id: UUID
    risk: str
    requested_at: datetime


@router.get("/approvals", response_model=list[PendingApprovalResponse])
async def list_approvals(request: Request, limit: int = 50) -> list[PendingApprovalResponse]:
    """Phase 4B -- the Approvals panel's initial state.

    Undecided only, oldest first. The panel then keeps itself current from
    `action.approval.requested` / `.decided` over `ws-gateway` rather than
    re-polling this: doc 04's "reads are pushed, never polled" applies to a
    queue whose whole purpose is to change while someone is looking at it.
    """
    approvals = await request.app.state.repository.list_pending_approvals(limit=limit)
    return [
        PendingApprovalResponse(
            action_id=approval.action_id,
            risk=approval.risk,
            requested_at=approval.requested_at,
        )
        for approval in approvals
    ]


class ApprovalDecisionRequest(BaseModel):
    approved: bool
    reason: str | None = None


class ApprovalDecisionResponse(BaseModel):
    action_id: UUID
    decision: str


@router.post("/approvals/{action_id}/decide")
async def decide_approval(
    action_id: UUID, body: ApprovalDecisionRequest, request: Request
) -> ApprovalDecisionResponse:
    repository = request.app.state.repository
    pending = await repository.find_pending_approval(action_id)
    if pending is None:
        raise HTTPException(status_code=404, detail="no pending approval for this action_id")
    if pending.decision is not None:
        raise HTTPException(status_code=409, detail="this approval has already been decided")
    decision = "approved" if body.approved else "denied"
    await repository.decide_pending_approval(
        action_id, decision=decision, decided_at=datetime.now(UTC)
    )
    return ApprovalDecisionResponse(action_id=action_id, decision=decision)

File: apps/web-client/src/entities/approvals.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { z } from "zod";

import type {
  ActionApprovalDecidedPayload,
  ActionApprovalRequestedPayload,
} from "@nova/nova-contracts";

import { gatewayFetch } from "./http";

/**
 * Actions waiting on a human decision.
 *
 * Two halves, and the split is the point. `GET /v1/action/approvals` gives
 * the queue as it stands when the panel opens; `action.approval.requested`
 * and `.decided` keep it current afterwards. A queue whose whole purpose is
 * to change while someone is looking at it is exactly what doc 04's "reads
 * are pushed, never polled" is about -- re-fetching on a timer would show a
 * request up to one interval late, which for an approval is the interval
 * during which NOVA is blocked.
 *
 * **No optimistic update.** Deciding an approval is shared cognitive state:
 * the row leaves this list when `action.approval.decided` says
 * action-engine recorded the decision, never when the button was clicked.
 * An optimistic removal would show a decision that a 409 (already decided
 * by someone else) or a network failure never actually made.
 */

const pendingApprovalSchema = z.object({
  action_id: z.string(),
  risk: z.string(),
  requested_at: z.string(),
});

export type PendingApproval = z.infer<typeof pendingApprovalSchema>;

const pendingApprovalsSchema = z.array(pendingApprovalSchema);

export const approvalKeys = { pending: ["approvals", "pending"] as const };

export function useApprovals() {
  return useQuery({
    queryKey: approvalKeys.pending,
    queryFn: async ({ signal }) => {
      return await gatewayFetch<PendingApproval[]>("/v1/action/approvals", pendingApprovalsSchema, { signal });
    },
  });
}

const decisionSchema = z.object({
  action_id: z.string(),
  decision: z.enum(["approved", "denied"]),
});

export function useDecideApproval() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({ actionId, approved }: { actionId: string; approved: boolean }) => {
      return await gatewayFetch(
        `/v1/action/approvals/${actionId}/decide`,
        decisionSchema,
        { method: "POST", body: { approved } },
      );
    },
    // Deliberately no `onMutate` optimistic removal, and no `setQueryData`
    // on success either: the row disappears when the bus confirms it. What
    // this does do is mark the query stale so a reconnect re-reads a queue
    // that may have moved on without us.
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: approvalKeys.pending, refetchType: "none" });
    },
  });
}

/** Fold a newly requested approval into the pending list, newest-last. */
export function reduceRequested(
  existing: PendingApproval[] | undefined,
  payload: ActionApprovalRequestedPayload,
): PendingApproval[] {
  const current = existing ?? [];
  if (current.some((approval) => approval.action_id === payload.action_id)) {
    return current;
  }
  const entry: PendingApproval = {
    action_id: payload.action_id,
    risk: payload.risk,
    requested_at: payload.requested_at,
  };
  // Oldest first, matching the endpoint: what has been blocked longest is
  // what most likely matters.
  return [...current, entry].sort((a, b) =>
    Date.parse(a.requested_at) - Date.parse(b.requested_at),
  );
}

/** Drop a decided approval. The queue answers "what is waiting", not "what happened". */
export function reduceDecided(
  existing: PendingApproval[] | undefined,
  payload: ActionApprovalDecidedPayload,
): PendingApproval[] {
  return (existing ?? []).filter((approval) => approval.action_id !== payload.action_id);
}
